server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeRole.java

```diff
@@ -172,6 +173,23 @@ public boolean isEnabledByDefault(final Settings settings) {
     public static final DiscoveryNodeRole MASTER_ROLE = new DiscoveryNodeRole("master", "m");
 
     public static final DiscoveryNodeRole REMOTE_CLUSTER_CLIENT_ROLE = new DiscoveryNodeRole("remote_cluster_client", "r");
+
+    public static final DiscoveryNodeRole ML_ROLE = new DiscoveryNodeRole("ml", "l");
+
+    public static final DiscoveryNodeRole TRANSFORM_ROLE = new DiscoveryNodeRole("transform", "t") {
+
+        @Override
+        public boolean isEnabledByDefault(final Settings settings) {
+            // transform is enabled by default on any non-frozen data node
+            final Predicate<DiscoveryNodeRole> notFrozen = Predicate.not(s -> Objects.equals(DiscoveryNodeRole.DATA_FROZEN_NODE_ROLE, s));
+            return DiscoveryNode.getPossibleRoles()
+                .stream()
+                .filter(notFrozen.and(DiscoveryNodeRole::canContainData))
+                .anyMatch(r -> DiscoveryNode.hasRole(settings, r));
+        }
```

Comment:
Now that this PR isn't being backported to 7.x I think this override is redundant as it boils down to `return true;` for the only situation where it's called, and the side benefit of keeping master and 7.x in sync doesn't apply if this code isn't going to be in 7.x. In 7.x it was necessary to keep the original logic of data nodes also being transform nodes unless otherwise specified, because users still have the option to use the deprecated `node.data: true`. If a user specifies _anything_ in `node.roles` then those are the exact roles, and defaults don't come into play, so for example `node.roles: [ master ]` means a node isn't a transform node even without this logic. I will open a followup PR to remove this, because it can confuse readers into thinking `node.roles: [ data ]` also makes a node a transform node.

Reply:
> I will open a followup PR to remove this

See #71455